src/attribute.cpp

```diff
@@ -27,6 +43,166 @@ Integer::Integer(uint8_t type, const uint8_t* data, size_t size)
+        m_value[i] = data[i];
+}
+
+IpAddress::IpAddress(uint8_t type, const std::array<uint8_t, 4>& address)
+    : Attribute(type),
+      m_value(address)
+{
+}
+
+std::vector<uint8_t> IpAddress::toVector(const std::string& secret, const std::array<uint8_t, 16>& auth) const
+{
+    std::vector<uint8_t> attribute(6);
+
+    attribute[0] = type();
+    attribute[1] = attribute.size();
+    for (size_t i = 0; i < m_value.size(); ++i)
+        attribute[i + 2] = m_value[i];
+
+    return attribute;
+}
+
+Encrypted::Encrypted(uint8_t type, const uint8_t* data, size_t size, const std::string& secret, const std::array<uint8_t, 16>& auth)
+        : Attribute(type)
+{
+    if (size > 128)
+        throw std::runtime_error("Invalid encrypted attribute size. Should be max 128 bytes, actual size is " + std::to_string(size));
+
+    std::vector<uint8_t> mdBuffer(16 + secret.length());
+
+    for (size_t i = 0; i < secret.length(); ++i)
+        mdBuffer[i] = secret[i];
+
+    for (size_t i = 0; i < 16; ++i)
+        mdBuffer[i + secret.length()] = auth[i];
+
+    std::vector<uint8_t> plaintext(size);
+
+    for (size_t i = 0; i < size / 16; ++i)
+    {
+        std::array<uint8_t, 16> md;
+
+        MD5(mdBuffer.data(), mdBuffer.size(), md.data());
+        for (size_t j = 0; j < 16; ++j)
+            plaintext[i * 16 + j] = data[i * 16 + j] ^ md[j];
+
+        for (size_t j = 0; j < secret.length(); ++j)
+            mdBuffer[j] = secret[j];
+
+        for (size_t j = 0; j < 16; ++j)
+            mdBuffer[j + secret.length()] = data[i * 16 + j];
+    }
+
+    m_value.assign(plaintext.begin(), plaintext.end());
+}
+
+Encrypted::Encrypted(uint8_t type, const std::string& password)
+    : Attribute(type),
+      m_value(password)
+{
+}
+
+std::vector<uint8_t> Encrypted::toVector(const std::string& secret, const std::array<uint8_t, 16>& auth) const
+{
+    std::string plaintext = m_value;
+
+    if (plaintext.length() % 16 != 0)
+    {
+        for (size_t i = 0; i < (16 - m_value.length() % 16); ++i)
+            plaintext.push_back(0);
+    }
+
+    std::vector<uint8_t> mdBuffer(16 + secret.length());
+
+    for (size_t i = 0; i < secret.length(); ++i)
+        mdBuffer[i] = secret[i];
+
+    for (size_t i = 0; i < 16; ++i)
+        mdBuffer[i + secret.length()] = auth[i];
+
+    std::vector<uint8_t> ciphertext(plaintext.length());
+
+    for (size_t i = 0; i < plaintext.length() / 16; ++i)
+    {
+        std::array<uint8_t, 16> md;
+
+        MD5(mdBuffer.data(), mdBuffer.size(), md.data());
+
+        for (size_t j = 0; j < 16; ++j)
+            ciphertext[i * 16 + j] = plaintext[i * 16 + j] ^ md[j];
+
+        for (size_t j = 0; j < secret.length(); ++j)
+            mdBuffer[j] = secret[j];
+
+        for (size_t j = 0; j < 16; ++j)
+            mdBuffer[j + secret.length()] = ciphertext[i * 16 + j];
+    }
+
+    auto it = ciphertext.begin();
+    it = ciphertext.insert(it, ciphertext.size() + 2);
+    it = ciphertext.insert(it, type());
```

Comment:
How about allocating `plaintext.length() + 2` bytes from the beginning, and filling the result with offset of 2?

Besides, you may use iterators instead of arithmetics:
```c++
std::vector<uint8_t> res(plaintext.length() + 2);
res[0] = type();
res[1] = res.size();
auto it = std::next(res.begin(), 2);

for (size_t i = 0; i < plaintext.length() / 16; ++i)
{
    // Calculate hash...
    
    for (size_t j = 0; j < md.size(); ++j)
        *it++ = plaintext[i * 16 + j] ^ md[j];
        
    // Fill mdBuffer...
}
```

You may also use another iterator for walking over plaintext. E.g.:
```c++
auto cit = std::next(res.begin(), 2);
auto pit = plaintext.begin();

// ...

*cit++ = *pit++ ^ md[j];

// ...
```

Although, I don't insist on this approach.